### PR TITLE
Various improvements in core HTTPConnection struct

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**
- redirects
- non-json responses
- custom content types

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
Any comments here @leandro-lucarella-sociomantic ?

File: source/octod/core.d [hunk 1]
```diff
@@ -19,6 +19,20 @@ import vibe.data.json;
...
+    bugs caused by typos in such string.
+ **/
+enum ResponseFormat
+{
+    /// JSON response for stable API methods
+    JsonStable  = "application/vnd.github.v3+json",
+    /// JSON response for in-development API methods
+    JsonPreview = "application/vnd.github.inertia-preview+json",
+    /// plain text string response with SHA
+    Version     = "application/vnd.github.VERSION.sha",
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d]: IIUC this has several errors:

1. Each change to GH API uses a different name, they are like different verions, is not like they always use `inertia-preview` for every preview, this string will always change
2. I think the way to ask for the response could have different parts, one is the version and another one is the format itself.
3. `Version` is not well formed, as `VERSION` should be replaced by a valid version, like `v3` or `inertia-preview`.

I will address this more deeply in the main review comment.

File: source/octod/core.d [hunk 2]
```diff
@@ -175,12 +175,14 @@ struct HTTPConnection
...
             url = GitHub API method URL (relative)
+            accept = optional argument with custom `Accept` header to
+                specify for this one request
 
         Returns:
             Json body of the response. If response is multi-page, all pages
             are collected and concatenated into one returned json object.
      **/
-    Json get ( string url )
+    Json get ( string url, string accept = "" )
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d]: Silly question, but isn't `null` a better representation for the lack of a value?
Also, I don't see where this new `accept` is being used :thinking:

File: source/octod/core.d [hunk 3]
```diff
@@ -206,7 +206,7 @@ struct HTTPConnection
                 (scope request) {
                     request.requestURL = url;
                     request.method = HTTPMethod.GET;
-                    this.prepareRequest(request);
+                    this.prepareRequest(request, accept);
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d]: Ha! Maybe this one belongs to the other commit (the add accept commit)?

File: source/octod/core.d [hunk 4]
```diff
@@ -305,12 +307,14 @@ struct HTTPConnection
             }
         );
 
-        this.handleResponseStatus(response);
+        if (auto location = this.handleResponseStatus(response))
+            return this.patch(location, json);
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d:325]: This changes seems to belong to another commit (Support HTTP 302 responses (redirects)).

File: source/octod/core.d [hunk 5]
```diff
@@ -217,6 +217,15 @@ struct HTTPConnection
...
             }
 
+            // Most responses are JSON, but there are also some plain
+            // text ones that have to be handled differently:
+
+            if (accept == ResponseFormat.Version)
+            {
+                import vibe.stream.operations;
+                return Json(response.bodyReader.readAllUTF8());
+            }
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d:212]: I think this is a bit more complex, there are other requests that are not json. Related to the design of how `accept` should be constructed.

**Review by leandro-lucarella-sociomantic — CHANGES_REQUESTED** [review]
So, there is a design choice I don't like about the accept string.

I think it could make more sense to have a function to build this string, because there could be a lot of combinations. Also, I don't think we should put *preview* versions as `enum`s, as they are fluid and will come and go. I think is reasonable to ask users that want to use preview features to write the version themselves.

The concrete format is `application/vnd.github[.version].param[+json]` (See https://developer.github.com/v3/media/#request-specific-version). Where `param` can vary with different objects, but it seems like the list is something like: raw, text, html, full, diff, patch, sha and base64 (only for gists apparently).

My suggestion then is something more like this:

```d
string getAccept(string version=ResponseApiVersion.V3,
        string format=ResponseFormat.Json,
        string param=ResponseParam.Default); // maybe different arguments order

enum ResponseApiVersion
{
    Current, // not specified
    V3,
}

enum ResponseFormat
{
    Default, // No +json
    Json,
}

enum ResponseParam
{
    Default, // not specified
    Raw,
    Text,
    Html,
    Full,
    Diff,
    Patch,
    Sha,
    Base64.
}
```

**Comment by mihails-strasuns-sociomantic** [comment]
@leandro-lucarella-sociomantic aha, I have missed the bit about well-defined format of `Accept` string, assumed it to be semi-magic. With your input I think best design would be to introduce `ResponseAccept` struct which can be initialized both from string and set of parameters, providing getter fields for specific metadata (like `format` and `version`)

**Comment by leandro-lucarella-sociomantic** [comment]
> @leandro-lucarella-sociomantic aha, I have missed the bit about well-defined format of Accept string, assumed it to be semi-magic. With your input I think best design would be to introduce ResponseAccept struct which can be initialized both from string and set of parameters, providing getter fields for specific metadata (like format and version)

OK, but I think you should always also be able to provide the accept string as a string, since this is pretty fluid, I think it would be a shame if a user of the library is limited and can't use some new param until the library gets updated.

File: source/octod/core.d (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/core.d]: In D there is no real difference between `null` and `""`, (same applies for `null` and empty array). `""` will have an address pointing to `\0` for compatibility with C strings but from the PoV of own D type system both are same arrays with `length == 0`. With that in mind I prefer `""` exactly because it leaves no confusion of `null` being special.

**Comment by mihails-strasuns-sociomantic** [comment]
Ok, I have completely changed how things work in 0a4e439, please check again.

After some thinking I have come to following conclusions:
- there is no point in configuring default media type for connection as correct one depends on specific request
- there is no point in configuring media type at all when higher level wrappers are used because those by design only work with one specific format yet know how to parse
- on core connection level it is necessary to allow user to specify custom media type for each request separately (to arbitrary media type)

API was adjusted accordingly. I have also created wrapper struct which can parse media types so that request code can branch depending on some subsets of that string (like checking if it is json or not).

**Comment by mihails-strasuns-sociomantic** [comment]
(will squash and reorder commits after review)

File: source/octod/core.d (hunk 2 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/core.d]: OK, but then there is a difference between `null` and `""` and you explicitly choose to use `""` to not care about which case is used. For me there is an important semantic difference between empty string and null (in set theory it would be the difference between an empty set and void). Is like a 3-state variable. Maybe in this case it's the same, but in other cases `""` could be a valid value and `null` the absence of a value (which could be turn in a default value that can't be expressed in compile-time for example).

In that sense, for me it makes more sense to use `null` to specify "I have no idea, just use whatever you think is best" (while `""` could mean please do really use `""` as the accept string).

**Comment by leandro-lucarella-sociomantic** [comment]
> there is no point in configuring default media type for connection as correct one depends on specific request

Mmm, I think using just `+json` makes sense, as probably 90% of the request will use that one.

File: source/octod/core.d [hunk 6]
```diff
@@ -218,9 +204,10 @@ struct HTTPConnection
             }
 
             // Most responses are JSON, but there are also some plain
-            // text ones that have to be handled differently:
+            // text ones that have to be handled differently (octod doesn't
+            // support XML responses):
 
-            if (accept == ResponseFormat.Version)
+            if (accept.format != "json")
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/core.d]: No enum for this "json"?

File: source/octod/media.d [hunk 7]
```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+
+    Wraps media type string used for interaction with GitHub as defined
+    by https://developer.github.com/v3/media
+
+    Copyright: Copyright (c) 2016 Sociomantic Labs. All rights reserved
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/media.d]: Missing `GmbH` :innocent:

File: source/octod/media.d [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+    /**
+        Constructor for constructing most common media type
+        directly (as opposed to string parsing)
+
+        Params:
+            format = format of response data (for example "json")
+            param = any custom parameters (for example "raw")
+     **/
+    this ( string format, string param = "" )
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/media.d:72]: Why `ver` can't be specified?

File: source/octod/media.d [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (!this.github_marker)
+            return std.format.format("application/%s", format);
+        else
+        {
+            return std.format.format(
+                "application/vnd.github%s%s%s",
+                this.ver.length ? "." ~ this.ver : "",
+                this.param.length ? "." ~ this.param : "",
+                this.format.length ? "+" ~ this.format : ""
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/media.d:109]: Fuck allocations?

File: source/octod/media.d [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        /**
+            Custom parameters, usually indicate kind of content to serve
+            (for example, rendered markdown or raw source)
+         **/
+        string param         = "";
+        /**
+            Format of response data, usually "json" or "xml"
+         **/
+        string format        = "json";
+    }
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/media.d:62]: If you are going the high-level route, I would define enum values for the currently existing`param`s and `format`s.

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
Just a few more comments.

BTW, are you going to fix the previous commits? I don't mind keeping the commits with the old details, but there are commits which have the diffs screwed with changes that belong to other commits. I think that's quite nasty. If you find it too complicated / annoying to fix them, I would even consider squashing all the commits...

File: source/octod/media.d (hunk 9 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:109]: Yes. Do you have problems with that? :)

File: source/octod/media.d (hunk 8 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:72]: Because any change of version implies revisiting all higher level wrappers. If one wants to use core connection utility directly, arbitrary string can be used instead.

File: source/octod/media.d (hunk 10 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:62]: AFAICS official spec doesn't limit possible legal values, that sounds like unnecessary limitation.

File: source/octod/core.d (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/core.d]: > OK, but then there is a difference between null and "" and you explicitly choose to use "" to not care about which case is used. 

D type system and runtime doesn't make any distinction between those two cases. Your interpretation of tristate would be a personal convention, quite a dangerous thing to rely on. For expressing tri-state in idiomatic manner one uses `std.typecons.Nullable!T`.

In D1 difference is more noticable because assigning null to array forces re-allocation instead of stomping. With D2 runtime it is not observable in reliable manner.

**Comment by mihails-strasuns-sociomantic** [comment]
> Mmm, I think using just +json makes sense, as probably 90% of the request will use that one.

Yeah but that is only applicable to direct usage of core connection. I think it is confusing to provide global configuration that will be ignore by major chunk of API.

> BTW, are you going to fix the previous commits?

Yes, after review.

File: source/octod/media.d (hunk 10 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/media.d:62]: I don't mean to only allow one value but just define the existing ones as "manifest constants". (yes, it was a great idea to reuse the `enum` keyword for manifest constants :trollface: )

File: source/octod/media.d (hunk 9 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/media.d:109]: Well, we all know how this ends up... Is harmless until some user cares about allocations. Would it be too hard to keep an internal buffer for formatting? Or this is too old school and nobody will ever expect a D2 library not to allocate like mad? :)

File: source/octod/media.d (hunk 8 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/media.d:72]: Mmm, I don't get it.

File: source/octod/media.d (hunk 9 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:109]: Your comment implies you greatly underestimate how bad octod currently behaves with allocations. Every single `format` call, every single JSON manipulation, big chunk of vibe.d utils - all this will allocate GC roots like crazy and attempt to fix it would result in full rewrite starting with HTTP facilities. This tiny piece of code is plain nothing compared to rest of allocations.

Considering this is a scripting purposed library, I have made an intentional decision to ignore allocation issues altogether - to be honest, I don't see much other options when using 3d-party upstream libraries.

File: source/octod/media.d (hunk 8 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:72]: This constructor is only intended to be used a shortcut to create required media type in higher level API (and changing version makes no sense there). When you use `connection.get(url, media)` directly, you can specify `media` as a plain string anyway.

File: source/octod/media.d (hunk 10 above)
Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:62]: I think it is just noise but don't mind either, will add.

Comment (mihails-strasuns-sociomantic) [reply source/octod/media.d:62]: Done

File: source/octod/media.d (hunk 9 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/media.d:109]: OK, that was my suspicion, that all the library ecosystem in D2 is so *broken* in terms of allocation that it is like trying to kill a whale with a flip-flop.

Point taken :+1:

File: source/octod/media.d (hunk 8 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/media.d:72]: My point was, why would you exclude core from using this facility? But is not important, never mind.

File: source/octod/media.d [hunk 11]
```diff
@@ -12,6 +12,23 @@
 module octod.media;
 
 /**
+    Enumerates some of commonly used formats
+ **/
+ enum MediaFormat
+ {
+     JSON = "json",
+     XML  = "xml",
+ }
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/media.d:21]: Params (raw, html, sha, etc.) missing. But I won't block this because of that anymore, we can do it in another PR.

**Review by leandro-lucarella-sociomantic — APPROVED** [review]

**Comment by leandro-lucarella-sociomantic** [comment]
Just waiting for the final touch ups to make it pretty enough to be merged (squashing/fixing commits).

**Comment by mihails-strasuns-sociomantic** [comment]
Rearranged commits

File: source/octod/core.d (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/core.d:325]: This change is still in the wrong commit.

File: source/octod/core.d (hunk 3 above)
Comment (leandro-lucarella-sociomantic) [reply source/octod/core.d]: And I think this one too.

File: source/octod/api/projects.d [hunk 12]
```diff
@@ -17,11 +17,18 @@
...
+import octod.media;
 import octod.api.common;
 
 import vibe.data.json;
 
 /**
+    Project API is experimental, thus relevant requests have to use
+    specific media type.
+ **/
+enum ProjectMediaType = "application/vnd.github.inertia-preview+json";
```
Comment (leandro-lucarella-sociomantic) [inline source/octod/api/projects.d:29]: Maybe is to address in another PR, but for me is weird that the user of the library can't decide if they are using a preview API or not. I have the feeling users should say somewhere which API version to use, and based on that certain objects/methods should be available or not.

For me is weird that using a non-stable API is hidden from the user.

**Comment by leandro-lucarella-sociomantic** [comment]
I still think the media type handling, more specifically the version, is not handled 100% correctly. I think the user should be able to control which version to use and based on that certain functionality should be available or not.

**Comment by mihails-strasuns-sociomantic** [comment]
Fixed misplaced changes in commits.

> I think the user should be able to control which version to use and based on that certain functionality should be available or not.

I disagree (with a medium level of violence :)). Higher level API wrappers are written with certain response format in mind thus they require specific version. When you use those, why would you ever care which version is used for one specific request if you only interact with wrapped responses?

And on core connection level you are in full control anyway.

**Comment by leandro-lucarella-sociomantic** [comment]
> I disagree (with a medium level of violence :)). Higher level API wrappers are written with certain response format in mind thus they require specific version. When you use those, why would you ever care which version is used for one specific request if you only interact with wrapped responses?

And I disagree (with a medium-high level of violence :D).

I agree that that the wrapper needs to know about the version used, what I don't agree is the user shouldn't know about it. I think there is a misunderstanding here. GitHub has a stable API (v3, which funny enough changes all the time (but only after a preview release was out for a long time), is like a stable `master` branch if you will, like a rolling release), and several unstable preview releases (the ones with funny names). Those come and go, and can have breaking changes often.

What I would expect from a high-level library is to only use the stable API, I don't want my code unexpectedly breaking without knowing I'm using an unstable API. Then, in an ideal world, I would like also to be able to **enable** experimental features if I need them, knowing they might break. What I envision as a library user is something like:

```d
auto api = new GitHubApi;
api.stableFeature();

auto api_preview = new GitHubApi(ApiVersion.InertiaPreview);
api_preview.stableFeature(); // OK
api_preview.previewFeature(); // OK
api.previewFeature(); // compile-time error.
```

How to achieve this, or if is the best D API for this, I'm not certain, I leave that to you. But do you see my point now?

> And on core connection level you are in full control anyway.

Yeah, but specially for being a high-level API, you shouldn't use preview API version without the user knowing.

**Comment by mihails-strasuns-sociomantic** [comment]
Would you be ok with simple boolean flag in config like `stable_only` (`false` by default). I don't want any explicit setting of API in higher level wrappers, it is waste of time and attention. As a I user I want to be able to read library docs and just call some methods, without caring about special configuration (especially if configuration results in nothing by protective enforcement, i.e. it either works or not)

**Comment by leandro-lucarella-sociomantic** [comment]
> Would you be ok with simple boolean flag in config like stable_only (false by default). I don't want any explicit setting of API in higher level wrappers, it is waste of time and attention.

I'm fine with it as a hack. As I am fine with it how it is nowbecause it's the infancy of the library. But I just wanted to stress out that for me this is very bad design. I wouldn't want to have a v1.0.0 tag without having an API similar to what I described in the [previous comment](https://github.com/sociomantic-tsunami/octod/pull/5#issuecomment-268264878).

BTW I think I would even prefer that for now we keep the hardcoded version in the module, because there is no point to waste time to make change stuff to still remain in the bad design side (specially when this library was done partially to use this new preview API :D). For now the library is unstable and experimental, but from now on it would be nice to think in terms of being able to chose which API version to use.

I'll create a new issue where we can discuss what direction we want to go with the design.

**Comment by leandro-lucarella-sociomantic** [comment]
https://github.com/sociomantic-tsunami/octod/issues/9

**Comment by mihails-strasuns** [comment]
> BTW I think I would even prefer that for now we keep the hardcoded version in the module

Hm, does that mean this is ready to be merged then or have I misunderstood?

**Comment by leandro-lucarella-sociomantic** [comment]
Yeah, I just needed to give a final look to all the commits :)
